server/manager/src/main/java/org/apache/accumulo/manager/MultipleManagerUtil.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.accumulo.manager;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.SortedSet;
+import java.util.TreeSet;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import org.apache.accumulo.core.data.TableId;
+import org.apache.accumulo.core.util.Pair;
+import org.apache.accumulo.server.ServerContext;
+
+public class MultipleManagerUtil {
+
+  /**
+   * Each Manager will be responsible for a range(s) of metadata tablets, but we don't want to split
+   * up a table's metadata tablets between managers as it will throw off the tablet balancing. If
+   * there are three managers, then we want to split up the metadata tablets roughly into thirds and
+   * have each manager responsible for one third, for example.
+   *
+   * @param context server context
+   * @param tables set of table ids
+   * @param numManagers number of managers
+   * @return list of num manager size, each element containing a set of tables for the manager to
+   *         manage
+   */
+  public static List<Set<TableId>> getTablesForManagers(ServerContext context, Set<TableId> tables,
+      int numManagers) {
+
+    if (numManagers == 0) {
+      throw new IllegalStateException("No managers, one or more expected");
+    }
+
+    if (numManagers == 1) {
+      return List.of(tables);
+    }
+
+    SortedSet<Pair<TableId,Long>> tableTabletCounts = new TreeSet<>(new Comparator<>() {
+      @Override
+      public int compare(Pair<TableId,Long> table1, Pair<TableId,Long> table2) {
+        // sort descending by number of tablets
+        int result = table1.getSecond().compareTo(table2.getSecond());
+        if (result == 0) {
+          return table1.getFirst().compareTo(table2.getFirst());
+        }
+        return -1 * result;
+      }
+    });
+    tables.forEach(tid -> {
```

Comment:
When the set of managers and tables are steady for a bit, all manager processes need to arrive at the same decisions for partitioning tables into buckets.  With the algorithm in this method different manager processes may see different counts for the same tables at different times and end up partitioning tables into different buckets. This could lead to overlap in the partitions or in the worst case a table that no manager processes.  We could start with a deterministic hash partitioning of tables and open a follow in issue to improve.  One possible way to improve would be to have a single manager process run this algorithm and publish the partitioning information, with all other manager just using it.

Reply:
Fluo partitions work among workers using an approach of having [one process determines the partitions](https://github.com/apache/fluo/blob/main/modules/core/src/main/java/org/apache/fluo/core/worker/finder/hash/PartitionManager.java) and puts those in ZK.  Then all other workers use those partitions from ZK to know what to work on.  This approach allows all of the workers to eventually settle on the same partitions which is what is needed here.  Posting the Fluo code to show that its not a lot of code and encapsulates nicely.  We could have a TabletManagementPartitioner that is created and tested as a stand alone task in its own PR that does this.